### PR TITLE
Add support for setting a FAT32 label

### PR DESCRIPTION
**Pull request by yadutaf**
This adds a ``SetLabel()`` to the ``FileSystem`` interface. On SquashFS and ISO 9660, this method returns an error, as the filesystems are read-only.

For FAT32, this new method re-uses code that was previously in ``Create()`` and ``Create()`` is updated to call this new method at filesystem initialization time. Doing this required to initialize the ``fs`` structure earlier, before writing anything to the destination. Let me know if this refactoring is acceptable!

On top of this, the PR seizes the occasion to move the label reading logic closer to Windows and dosfstools' ``fatlabel`` behavior. That is, they read the label from the root directory rather than the bootsector. And, according to ``fatlabel``'s man page, Windows does not even write the label to the bootsector, which would cause go-diskfs to fail detecting it.

For the tests, I manually modified the fixture with "fatlabel" to include a label by default and updated the Readme. Was this the proper way of doing ?

### CODE REVIEW

**Comment by deitch** [comment]
"I torture packets / containers."

😆

**Comment by deitch** [comment]
Adding `SetLabel()` makes sense, and the returns from read-only filesystems are correct.

I don't mind terribly moving all of the "write to disk" from `fat32.Create()` to the end, as opposed to interspersed. Moving just one of them to its own function `writeBootSector()` but not the others - e.g. `writeFsis`, `writeFat` - is a bit unbalanced.

Having `SetLabel()` means you need to add the enforcement of the label format into there, so might as well take advantage and use the raw label in the `dos71EBPB` structure, although that would require changing the comment [here](https://github.com/diskfs/go-diskfs/blob/d704acd18db2dae47d0cbb2cbed5850016b13fc5/filesystem/fat32/dos71bpb.go#L51). 

> That is, they read the label from the root directory rather than the bootsector. And, according to fatlabel's man page

I am somewhat torn on this. For `SetLabel()`, I like what you did, i.e. write it to both places. For read, are there compatibility issues? I cannot recall the driver behind the original implementation, but I am pretty sure the logic was taken from how Linux does it. Do we run the risk that some tool will write it _only_ to the boot sector volume label, and this `Label()` read code will miss it? Is there some other construct that we can use to distinguish between _on-filesystem_ label and _on bootsector_ label?

> For the tests, I manually modified the fixture with "fatlabel" to include a label by default and updated the Readme. Was this the proper way of doing ?

Unfortunately, yes. I have not come up with a better way.

File: filesystem/fat32/fat32.go [hunk 1]
```diff
@@ -562,14 +571,87 @@ func (fs *FileSystem) OpenFile(p string, flag int) (filesystem.File, error) {
 	}, nil
 }
 
-// Label get the label of the filesystem
+// Label get the label of the filesystem from the secial file in the root directory.
+// The label stored in the boot sector is ignored to mimic Windows behavior which
+// only stores and reads the label from the special file in the root directory.
 func (fs *FileSystem) Label() string {
-	// be sane about everything existing
+	// locate the filesystem root directory
```
Comment (deitch) [inline filesystem/fat32/fat32.go:578]: All of this reads _solely_ from the volume label magic file on the filesystem, and drops reading the volume label on the boot sector. 

Is there a compatibility issue? Are there implementations that write solely there, and we will miss it? Should we read from magical file and fall back to volume label on boot sector, or is that wrong? Conversely, is there a separate construct for that label that we could use?

**Review by deitch — CHANGES_REQUESTED** [review]
A few small changes. This is pretty nice.

File: filesystem/fat32/fat32.go [hunk 2]
```diff
@@ -307,6 +251,32 @@ func Create(f util.File, size, start, blocksize int64, volumeLabel string) (*Fil
 		file:            f,
 	}
 
+	// write the boot section
+	if err := fs.writeBootSector(); err != nil {
+		return nil, fmt.Errorf("failed to write the boot sector")
+	}
+
+	// write the fsis
```
[the remark this answers is not in the document]
Comment (yadutaf) [reply filesystem/fat32/fat32.go:259]: Done and integrated in `allocateSpace`

File: filesystem/fat32/fat32.go [hunk 3]
```diff
@@ -199,7 +194,7 @@ func Create(f util.File, size, start, blocksize int64, volumeLabel string) (*Fil
 		bootFileName:          [12]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
 		extendedBootSignature: longDos71EBPB,
 		volumeSerialNumber:    volid,
-		volumeLabel:           fmt.Sprintf("%-11.11s", volumeLabel), // "NO NAME    "
+		volumeLabel:           volumeLabel, // "NO NAME    "
```
[the remark this answers is not in the document]
Comment (yadutaf) [reply filesystem/fat32/fat32.go]: Thinking of it, the enforcement should be preserved to remain compatible with the specification and thus the comment should not change. Instead, I need to fix this line. A changed has been pushed to initialize this field to "NO NAME". The size enforcement is done in ``SetLabel()``.

**Comment by yadutaf** [comment]
Thanks for the review! I like the suggestion of introducing ``writeFsis`` and ``writeFat`` in the same way as ``writeBootSector``; this is also the occasion to share more code with ``allocateSpace``. I pushed a new proposition with these changes in a dedicated refactoring commit.

Regarding the compatibility concern, I might be a little bit too optimistic here, but I was considering that Windows is the reference implementation while Linux is the main alternative implementation. On Linux, the kernel does not read nor writes the label. There was a proposition some times ago to introduce an `ioctl` for it, but it did not get merged. Thus, the reference Linux implementation is ``fatlabel`` and I followed carefully the advertised behavior.

If it proves to be problematic, future options could be:
* Add a fallback path on the bootsector, but we may end up with stranger consistency issues.
* Add a FAT32 specific helper to query the label from the boot sector.

Would it be acceptable this way ?

**Comment by deitch** [comment]
So are you saying that:

1. Windows is the reference, which you implemented here in this PR.
2. Linux does not do anything with it in the kernel at all.
3. Linux does it in `fatlabel` (userspace), which follows your implementation.

If that is correct, then I think we can work with this. Going to take another look at the changes.

**Comment by deitch** [comment]
Going to kick off CI. If it all passes, are you good to go?

**Review by deitch — DISMISSED** [review]

**Comment by yadutaf** [comment]
Oops, the lint. Sorry... Will push a fix this evening. Aside from this, good to.

**Review by deitch — APPROVED** [review]

**Pull request merged**